mediatek: add support for Zbtlink ZBT-Z8103AX

darknblack opened the pull request:
Specifications:
SoC: MediaTek MT7981B
RAM: 256MiB
Flash: SPI-NAND 128 MiB
Switch: 1 WAN, 3 LAN (Gigabit)
Buttons: Reset, Mesh
Power: DC 12V 1A
WiFi: MT7976CN
UART: 115200n8
UART Layout:
VCC-RX-TX-GND

No. of Antennas: 6
Note: Upon opening the router, I was surprised to find out that only
5 antennas were actually connected to the mainboard.

Led Layout:
Power-Mesh-WLAN2G-WAN-LAN3-LAN2-LAN1-WLAN5G

Buttons:
Reset-Mesh

Installation:
A. Through OpenWrt Dashboard:
If your router comes with OpenWrt preinstalled (modified by the seller), you can easily upgrade by going to the dashboard (192.168.1.1) and then navigate to System -> Backup/Flash firmware, then flash the firmware

B. Through TFTP
Standard installation via UART:

1. Connect USB Serial Adapter to the UART, (NOTE: Don't connect the VCC pin).
2. Power on the router. Make sure that you can access your router via UART.
3. Restart the router then repeatedly press ctrl + c to skip default boot.
4. Type > bootmenu
5. Press '2' to select upgrade firmware
6. Press 'Y' on 'Run image after upgrading?'
7. Press '0' and hit 'enter' to select TFTP client (default)
8. Fill the U-Boot's IP address and TFTP server's IP address.
9. Finally, enter the 'firmware' filename.

Everything seems to work fine except for unstability when the `country code` is set to `US` and the `maximum transmit power` is set to more than 23dBm.

dangowrt commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,271 @@
+		port@0 {
+			reg = <0>;
+			label = "lan1";
+			nvmem-cell-names = "mac-address";
```
port 0~3 don't need mac-address form NVMEM, they will inhert it from gmac0 where it is already set.

dangowrt commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,271 @@
+			label = "wan";
+			nvmem-cell-names = "mac-address";
+			nvmem-cells = <&macaddr_factory_4>;
+			mac-address-increment = <1>;
```
`mac-address-increment` has been deprecated.
See for example https://github.com/openwrt/openwrt/pull/13956/commits/c6d05de704aa8ba72073089eec4a788dd8bc70e8 on how to replace it with `mac-base`.

dangowrt commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -110,6 +111,11 @@ mediatek_setup_macs()
 		wan_mac=$label_mac
 		lan_mac=$(macaddr_add "$label_mac" 2)
 		;;
+	zbtlink,zbt-z8103ax)
```
You are already setting up MAC addresses in device tree, no need to do it again in userspace. Remove it here.

csharper2005 replied:
sorting

dangowrt reviewed (changes requested):
Looks good, just some redundant assignments can (and should) be removed.

csharper2005 commented on package/boot/uboot-envtools/files/mediatek_filogic:
```diff
@@ -65,6 +65,7 @@ mercusys,mr90x-v1)
 	local envdev=/dev/mtd$(find_mtd_index "u-boot-env")
 	ubootenv_add_uci_config "$envdev" "0x0" "0x20000" "0x20000" "1"
 	;;
+zbtlink,zbt-z8103ax|\
```
alphabetical sorting

csharper2005 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,271 @@
+
+		net {
+			label = "blue:network";
+			gpios = <&pio 8 GPIO_ACTIVE_LOW>;
```
Would be nice to have leds and buttons sorted by the their gpio numbers

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -9,6 +9,12 @@ case $board in
 cudy,wr3000-v1)
 	ucidef_set_led_netdev "wan" "wan" "blue:wan" "wan"
 	;;
+zbtlink,zbt-z8103ax)
```
sorting

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -32,6 +32,7 @@ mediatek_setup_interfaces()
 	glinet,gl-mt3000)
 		ucidef_set_interfaces_lan_wan eth1 eth0
 		;;
+	zbtlink,zbt-z8103ax|\
```
sorting

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -755,3 +755,20 @@ define Device/zyxel_nwa50ax-pro
 TARGET_DEVICES += zyxel_nwa50ax-pro
+
+define Device/zbtlink_zbt-z8103ax
+  DEVICE_VENDOR := ZBT
```
Why `ZBT`? `zbtlink_zbt-z8103ax` means vendor `zbtlink`, model `zbt-z8103ax`

darknblack commented:
Hi! This is my first pull request, and my first time contributing to an open source project.
I've updated the branch and addressed the feedback from @csharper2005 and @dangowrt. 

Thanks for your feedback and help. I really appreciate it. :v:

csharper2005 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,262 @@
+#include "mt7981.dtsi"
+
+/ {
+	model = "ZBT Z8103AX";
```
Why not `Zbtlink ZBT-Z8103AX`? Probably the device model in the commit title ("[mediatek: add support for ZBT Z8103AX](https://github.com/openwrt/openwrt/pull/13933/commits/936eebdd11b6cdf10c68374d9b0b9e004c4059c7)") also should be fixed.

darknblack replied:
I think ZBT Z8103AX will do for the 'model'. I based it on the existing routers made by Zbtlink.
https://github.com/openwrt/openwrt/commit/046263095e007ba80bdf3525de32d7cce11734e8

csharper2005 replied:
Hmm... Indeed, we have 1 ZBT and 21 Zbtlink https://github.com/search?q=repo%3Aopenwrt%2Fopenwrt+model+%3D+"ZBT&type=code&p=1. Nevertheless, I suppose we need either ZBT or Zbtlink in one commit, not mix.

csharper2005 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,262 @@
+
+		mesh {
+			label = "mesh";
+			gpios = <&pio 0 GPIO_ACTIVE_HIGH>;
```
The buttons also can ne sorted.

csharper2005 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,262 @@
+&wifi {
+	status = "okay";
+	mediatek,mtd-eeprom = <&factory 0x0>;
+};
```
Missing newline at the end of the file.

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -29,6 +29,12 @@ xiaomi,redmi-router-ax6000-stock|\
+zbtlink,zbt-z8103ax)
+	ucidef_set_led_netdev "lan1" "lan1" "green:status" "lan1" "link tx rx"
+	ucidef_set_led_netdev "lan2" "lan2" "green:status" "lan2" "link tx rx"
+	ucidef_set_led_netdev "lan3" "lan3" "green:status" "lan3" "link tx rx"
```
Why LAN 1-3 is present here but LAN4 not?

csharper2005 replied:
> The router has the following physical ports:
> 
> * Lan1
> * Lan2
> * Lan3
> * Wan

Ok. My suggestion is to drop lan4 from dts in this case,

> I based my DTS file here: https://forum.openwrt.org/t/zbt-z8103ax-d-upgrading-from-21-02-to-23-05/174582/7

Oh no, this is a dts from ROOTer. lol I thought that ROOTer is made for devices with usb port only.

csharper2005 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,262 @@
+				reg = <0x100000 0x80000>;
+			};
+
+			factory: partition@180000 {
```
Can the `factory:` label be dropped?

darknblack replied:
I'm new to these kind of stuff but how do we handle the following without the `factory` label?
`
&wifi {
	status = "okay";
	mediatek,mtd-eeprom = <&factory 0x0>;
};
`

I based it merely on the existing DTS files from devices with the same processor (MT7981).

csharper2005 replied:
You're right. You use old-style and the label should be kept.

New-style example:
https://github.com/openwrt/openwrt/pull/13947/files#diff-412a2b33d28362f746588ae3f52f1029aebdad56b29c7b734233866bf0258a90R89
https://github.com/openwrt/openwrt/pull/13947/files#diff-1a5ad50d6a2eac8c748755abcce1dd94df6b0e6f48dfbb917341b8c29ebd82b3R14

darknblack replied:
Btw, is there a benefit in using the new-style?

raenye replied:
IMHO new devices submitted should use new style and avoid deprecated syntax, even if it still works.

csharper2005 reviewed (commented):
@darknblack thanks for your updates. Are the all MAC addresses in OpenWrt equal to the stock firmware one?

Do you know how to squash the commits into the one?

darknblack commented:
>Are the all MAC addresses in OpenWrt equal to the stock firmware one?
Yes they do now.

> Do you know how to squash the commits into the one?
I think I can but i'm not 100% sure, I really appreciate if you can help me on this one.

csharper2005 commented:
> I think I can but i'm not 100% sure, I really appreciate if you can help me on this one.

`git rebase -i HEAD~20`
edit and save:
![image](https://github.com/openwrt/openwrt/assets/82600916/ede794ee-e6f8-49d0-a476-504bd58dfa4a)
`git push --force`

darknblack commented:
I applied all of your recommendations to the latest commit. Thanks for the help @csharper2005 

I learned a lot today :bow: :raised_hands:

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -755,3 +755,20 @@ define Device/zyxel_nwa50ax-pro
 endef
 TARGET_DEVICES += zyxel_nwa50ax-pro
+
+define Device/zbtlink_zbt-z8103ax
```
Sorting.

csharper2005 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,257 @@
+		port@3 {
+			reg = <3>;
+			label = "wan";
+		};
```
Why `wan`? According to your `02_network`, the `wan` is `eth1`. `port@3` should be dropped in this case.

csharper2005 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,257 @@
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>; /* 0x198  */
+			gpios = <&pio 1 GPIO_ACTIVE_LOW>; /* GPIO_ACTIVE_HIGH = 0x00 | GPIO_ACTIVE_LOW = 0x01 */
```
Long line. Ideally should be < 75 characters. For example:
```
/* GPIO_ACTIVE_HIGH = 0x00 | GPIO_ACTIVE_LOW = 0x01 */
gpios = <&pio 1 GPIO_ACTIVE_LOW>;
```

csharper2005 reviewed (commented):
And still commit title. Usually it looks like that
`mediatek: add support for device_vendor device_model`

csharper2005 commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,253 @@
+			label = "lan3";
+		};
+
+
```
This line (one of the two) can be dropped.

And you probably should add a blank line between commit title and commit message. Otherwise we have such a mess:
![image](https://github.com/openwrt/openwrt/assets/82600916/962fec5a-15a4-4cf1-bdd0-fc295aee246a)

darknblack commented:
Hi, I've made several changes to the branch based on the comments and feedback from @csharper2005 and @dangowrt:
- Removed some comments from the DTS file.
- Updated the "blue:network" LED label to "green:network".
- Added a blank line between the commit title and message.
- Included the "Zbtlink" vendor name in both the commit message and the pull request title.
- Sorting of props.

I flashed the latest firmware on my router and i'm glad that there's no more unnecessary interface devices :ok_man: :crossed_fingers:

Thanks for the help. :relaxed:

csharper2005 reviewed (approved):
LGTM. The only thing I would suggest thinking about is `nvmem-layout` for eeprom definition instead of the old-style.

darknblack commented:
> LGTM. The only thing I would suggest thinking about is `nvmem-layout` for eeprom definition instead of the old-style.

Hi, how do we get the correct eeprom size (reg)?

darknblack commented:
> > LGTM. The only thing I would suggest thinking about is `nvmem-layout` for eeprom definition instead of the old-style.
> 
> Hi, how do we get the correct eeprom size (reg)?

Btw here's the content of the `Factory`

```
root@OpenWrt:~# hexdump -C /dev/mtd2
00000000  81 79 00 00 f8 5e 3c 4f  08 ca f8 5e 3c 4f 08 cb  |.y...^<O...^<O..|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000020  00 00 00 00 f8 5e 3c 4f  08 cd f8 5e 3c 4f 08 cc  |.....^<O...^<O..|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000190  12 5b 48 4c 00 28 00 00  05 d0 00 00 00 00 00 00  |.[HL.(..........|
000001a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000240  00 00 00 00 00 00 00 00  00 00 00 00 a2 a1 a2 86  |................|
00000250  a3 91 00 91 00 91 00 91  00 91 00 00 00 00 00 00  |................|
00000260  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000270  0c 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000280  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000360  00 00 f5 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000370  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000003e0  00 00 00 00 00 00 00 12  12 12 00 00 00 00 00 00  |................|
000003f0  00 00 00 00 00 00 22 22  22 22 33 33 33 33 33 33  |......""""333333|
00000400  33 33 33 33 33 33 33 33  33 33 33 33 33 33 33 33  |3333333333333333|
00000410  33 33 33 33 33 33 00 00  00 00 00 00 00 00 00 00  |333333..........|
00000420  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000440  00 24 24 29 29 24 24 24  24 24 24 24 24 24 24 24  |.$$))$$$$$$$$$$$|
00000450  24 24 24 24 28 28 28 28  28 00 00 00 00 00 00 00  |$$$$(((((.......|
00000460  00 00 00 00 00 24 24 24  24 22 22 22 22 24 24 24  |.....$$$$""""$$$|
00000470  24 22 22 22 22 24 24 24  24 22 22 22 22 24 24 24  |$""""$$$$""""$$$|
00000480  24 22 22 22 22 00 00 00  00 00 00 00 00 00 00 00  |$""""...........|
00000490  00 00 00 00 00 c4 c4 c4  c4 c4 c4 00 00 00 00 00  |................|
000004a0  00 c4 c4 c4 c4 c6 c4 c6  c8 c7 c6 c4 c4 c4 c4 c6  |................|
000004b0  c4 c4 c6 c7 c7 c4 c4 c3  c2 c3 c2 c3 c5 c4 c5 00  |................|
000004c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000540  00 00 00 00 00 00 00 00  00 00 44 00 00 00 00 00  |..........D.....|
00000550  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000580  00 00 00 00 00 7f 7f 7f  d1 d1 dd dd e9 e9 f5 f5  |................|
00000590  01 01 09 09 15 15 d1 d1  dd dd e9 e9 f5 f5 fd fd  |................|
000005a0  09 09 15 15 d1 d1 dd dd  e9 e9 f5 f5 fd fd 09 09  |................|
000005b0  15 15 d1 d1 dd dd e9 e9  f5 f5 fd fd 09 09 15 15  |................|
000005c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000600  1c f2 fc 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000610  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000790  02 1e 02 1e 02 00 02 00  02 37 02 37 02 37 02 37  |.........7.7.7.7|
000007a0  02 37 02 37 02 37 02 37  02 37 02 37 02 37 02 37  |.7.7.7.7.7.7.7.7|
*
000007d0  00 22 00 c6 c6 c4 c4 c3  00 00 c1 c2 c1 83 83 83  |."..............|
000007e0  83 86 86 c1 c1 83 83 83  83 86 86 c2 c1 c1 81 81  |................|
000007f0  82 83 85 85 86 86 c1 c1  81 81 82 83 85 85 86 86  |................|
00000800  c1 c1 81 81 82 83 85 85  86 86 c1 c1 81 81 82 83  |................|
00000810  85 85 86 86 c1 c1 81 81  82 83 85 85 86 86 c5 c5  |................|
00000820  c3 c1 00 c2 c3 c2 c2 00  81 82 83 83 c2 c2 c2 00  |................|
00000830  81 82 83 83 c3 c1 c1 c1  81 82 83 84 84 c2 c2 c2  |................|
00000840  c2 00 81 82 83 83 85 85  c2 c2 c2 00 81 82 83 83  |................|
00000850  85 85 c1 c1 c1 81 82 83  84 84 86 86 82 82 84 84  |................|
00000860  85 86 87 87 89 89 c2 c2  c2 00 81 82 83 83 85 85  |................|
00000870  c2 c2 c2 00 81 82 83 83  85 85 c2 c2 c2 00 81 82  |................|
00000880  83 83 85 85 c4 c4 c2 c1  00 c3 c3 c3 c1 00 81 83  |................|
00000890  83 86 86 c3 c3 c3 c1 00  81 83 83 86 86 c2 c2 c2  |................|
000008a0  00 81 82 84 84 86 86 82  82 82 84 85 86 88 88 8b  |................|
000008b0  8b c3 c3 c3 c1 00 81 83  83 86 86 c3 c3 c3 c1 00  |................|
000008c0  81 83 83 86 86 c3 c3 c3  c1 00 81 83 83 86 86 00  |................|
000008d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000990  00 b3 00 00 00 b3 00 00  00 82 00 00 00 00 00 00  |................|
000009a0  01 00 00 00 00 00 c4 00  b9 c5 88 00 00 00 00 00  |................|
000009b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000a20  00 49 52 56 55 55 55 55  00 4a 52 51 51 51 51 51  |.IRVUUUU.JRQQQQQ|
00000a30  00 4a 53 55 56 56 56 56  00 49 50 4e 51 51 51 51  |.JSUVVVV.IPNQQQQ|
00000a40  00 49 51 50 54 54 54 54  00 49 50 51 51 51 51 51  |.IQPTTTT.IPQQQQQ|
00000a50  00 49 52 51 50 50 50 50  00 49 52 51 51 51 51 51  |.IRQPPPP.IRQQQQQ|
00000a60  00 49 52 51 54 54 54 54  00 49 51 50 54 54 54 54  |.IRQTTTT.IQPTTTT|
00000a70  00 49 53 52 51 51 51 51  00 49 53 53 52 52 52 52  |.ISRQQQQ.ISSRRRR|
00000a80  00 49 51 50 50 50 50 50  00 49 51 52 54 54 54 54  |.IQPPPPP.IQRTTTT|
00000a90  00 49 52 51 53 53 53 53  00 00 00 00 00 00 00 00  |.IRQSSSS........|
00000aa0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000b10  d1 d1 dd dd e9 e9 f5 f5  fd fd 14 14 1d 1d 00 00  |................|
00000b20  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000ca0  83 84 00 00 c3 c4 c3 82  82 82 81 c1 c4 c5 c4 00  |................|
00000cb0  00 00 c1 c3 c4 c4 c4 81  81 81 c1 c2 00 00 00 00  |................|
00000cc0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000d00  c0 bb c0 bb c0 bb c0 bb  40 c5 c0 c4 c0 c3 c0 c3  |........@.......|
00000d10  40 c3 40 c5 40 c4 c0 c3  40 c3 c0 c2 c0 c5 c0 c4  |@.@.@...@.......|
00000d20  40 c4 40 c4 c0 c3 c0 c5  c0 c4 40 c4 40 c4 c0 c3  |@.@.......@.@...|
00000d30  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000d40  00 00 00 00 00 00 00 00  40 c6 40 c6 40 c6 40 c6  |........@.@.@.@.|
00000d50  40 c6 40 c6 40 c6 40 c6  40 c6 40 c6 40 c6 40 c6  |@.@.@.@.@.@.@.@.|
*
00000d80  40 c6 40 c6 40 c6 40 c6  00 00 00 00 00 00 00 00  |@.@.@.@.........|
00000d90  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00001000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00008000  ff ff ff ff 78 a3 51 22  33 55 ff ff ff ff ff ff  |....x.Q"3U......|
00008010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00200000
```

cc: @csharper2005

dangowrt commented:
EEPROM size for MT7981 is like MT7916:
```
mt7915/mt7915.h:#define MT7916_EEPROM_SIZE		4096
```
so `<0x0 0x1000>` is the correct nvmem cell definition.

raenye commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,273 @@
+	gpio-keys {
+		compatible = "gpio-keys";
+
+		mesh {
```
```suggestion
		button-mesh {
```

raenye commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,273 @@
+			gpios = <&pio 0 GPIO_ACTIVE_HIGH>;
+		};
+		
+		reset {
```
```suggestion
		button-reset {
```

raenye commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,273 @@
+			gpios = <&pio 8 GPIO_ACTIVE_LOW>;
+		};
+
+		blue {
+			label = "blue:status";
```
```suggestion
		led-status-blue {
			color = <LED_COLOR_ID_BLUE>;
			function = LED_FUNCTION_STATUS;
```

raenye commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,273 @@
+			gpios = <&pio 23 GPIO_ACTIVE_LOW>;
+		};
+
+		led_status_green: green {
+			label = "green:status";
```
```suggestion
		led_status_green: led-status-green {
			label = "green:status"; // this needs to stay until #13837 is merged
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_STATUS;
```

raenye commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,273 @@
+			gpios = <&pio 24 GPIO_ACTIVE_LOW>;
+		};
+		
+		led_status_red: red {
+			label = "red:status";
```
```suggestion
		led_status_red: led-status-red {
			label = "red:status"; // this needs to stay until #13837 is merged
			color = <LED_COLOR_ID_RED>;
			function = LED_FUNCTION_STATUS;
```

raenye commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,273 @@
+	leds {
+		compatible = "gpio-leds";
+
+		net {
+			label = "green:network";
```
Network is kinda general, but I see it is used for WAN below.

```suggestion
		led-wan {
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_WAN;
```

raenye commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,273 @@
+};
+
+&mdio_bus {
+	switch: switch@0 {
+		compatible = "mediatek,mt7531";
+		reg = <31>;
```
Given `reg = <31>`, shouldn't it be `switch@31`?

raenye commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -29,6 +29,12 @@ xiaomi,redmi-router-ax6000-stock|\
+	ucidef_set_led_netdev "lan1" "lan1" "green:status" "lan1" "link tx rx"
+	ucidef_set_led_netdev "lan2" "lan2" "green:status" "lan2" "link tx rx"
+	ucidef_set_led_netdev "lan3" "lan3" "green:status" "lan3" "link tx rx"
+	ucidef_set_led_netdev "wan" "wan" "green:network" "eth1" "link tx rx"
```
```suggestion
	ucidef_set_led_netdev "wan" "wan" "green:wan" "eth1" "link tx rx"
```

raenye commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -95,6 +95,11 @@ case "$board" in
 	tplink,tl-xdr6088)
 		[ "$PHYNBR" = "0" ] && get_mac_label > /sys${DEVPATH}/macaddress
 		;;
+	zbtlink,zbt-z8103ax)
+		addr=$(mtd_get_mac_binary "Factory" 0x4)
+		[ "$PHYNBR" = "0" ] && macaddr_add $addr 2 > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" = "1" ] && macaddr_add $addr 3 > /sys${DEVPATH}/macaddress
+		;;
```
Is this needed? both PHYs have MAC address assigned via DTS.

darknblack replied:
Hi @raenye, I had to add this again because I can't make the DTS one work without issue.

raenye replied:
Neither phy worked? or just phy0? I guess the problem is that MT7976CN is a DBDC IC.

dangowrt replied:
FYI https://patchwork.kernel.org/project/linux-wireless/list/?series=771380

raenye replied:
@dangowrt cool! Until upstream accepts, is there a patch to mt76 within OpenWrt?

Also, is this only for mt7976? sounds like mt7615dn could benefit from it too.

raenye commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -728,6 +728,23 @@ endif
+  PAGESIZE := 2048
+  IMAGE_SIZE := 65536k
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-ubi | check-size $$$$(IMAGE_SIZE)
```
I think double-escaping is not needed here.
```suggestion
  IMAGE/factory.bin := append-ubi | check-size $$(IMAGE_SIZE)
```

darknblack replied:
Hi @raenye, for the sake of the consistency within the file, I'll keep it like this. Maybe this is better handled on a separate commit.

raenye reviewed (commented):
Good work, here are my suggestions.
It'd be nice to mention in the commit message all buttons and LEDs (including non-GPIO ones that are not in the dts), as well as the number of antennas (helps identifying the device).

raenye commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -29,6 +29,12 @@ xiaomi,redmi-router-ax6000-stock|\
 	ucidef_set_led_netdev "wan" "wan" "rgb:network" "wan"
 	;;
+zbtlink,zbt-z8103ax)
+	ucidef_set_led_netdev "lan1" "lan1" "green:status" "lan1" "link tx rx"
+	ucidef_set_led_netdev "lan2" "lan2" "green:status" "lan2" "link tx rx"
+	ucidef_set_led_netdev "lan3" "lan3" "green:status" "lan3" "link tx rx"
```
Does this (3 devices using the same LED) really work at expected? I'd expect the last one to prevail.

csharper2005 replied:
As idea: to bind this led to "br-lan" instead of the lan ports. @darknblack what is indicating led in stock firmware?

darknblack replied:
Hi, it turns out, I don't have to add lan1-3 in the file. It seems like the led will automatically light up when you insert an rj45 to the port.

Also, I have added wlan2g and wlan5g. Thank you :)

darknblack commented:
Hi, 
I made a bunch of updates including:
 - Added wlan2g and wlan5g led config
 - Removed unnecessary lan1-lan3 led config
 - Updated the led naming convention

Thanks for the help @raenye 🙆‍♂️ 🤞

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -29,6 +29,11 @@ xiaomi,redmi-router-ax6000-stock|\
+zbtlink,zbt-z8103ax)
+	ucidef_set_led_netdev "wan" "wan" "green:wan" "eth1" "link tx rx"
+	ucidef_set_led_netdev "wlan2g" "wlan2g" "green:wlan2g" "phy0-ap0"
+	ucidef_set_led_netdev "wlan5g" "wlan5g" "green:wlan5g" "phy1-ap0"
```
You can use dts `linux,default-trigger` for the wireless leds. The dts is more preferable way than the script. Here is an example:
https://github.com/openwrt/openwrt/blob/main/target/linux/ramips/dts/mt7621_tplink_ec330-g5u-v1.dts#L76
https://github.com/openwrt/openwrt/blob/main/target/linux/ramips/dts/mt7621_tplink_ec330-g5u-v1.dts#L85

darknblack replied:
Ohhh, but what is the correct value should I put to it? 
Will this work? `linux,default-trigger = "phy0-ap0";`

csharper2005 replied:
Please try `linux,default-trigger = "phy0tpt";` for phy0 and `linux,default-trigger = "phy1tpt";` for phy1 (like in example).

darknblack replied:
Thank you it worked :bow: :partying_face:

hauke commented:
@dangowrt Just added support for the ZBT-Z8102AX, please rebase this PR.

dangowrt commented on package/boot/uboot-envtools/files/mediatek_filogic:
```diff
@@ -56,6 +56,7 @@ cmcc,rax3000m)
 	;;
 cetron,ct3003|\
 netgear,wax220|\
+zbtlink,zbt-z8103ax|\
```
Please maintain alphabetic order, ie. put z8103ax after z8102ax.

darknblack replied:
Yes! Sorry, I was having a hard time rebasing. Thank you✌️🤞

dangowrt commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,273 @@
+				reg = <0x180000 0x200000>;
+				read-only;
+
+				compatible = "nvmem-cells";
```
```suggestion
```

darknblack replied:
Removed. Thanks 👌

darknblack commented:
> @dangowrt Just added support for the ZBT-Z8102AX in [c8c2f52](https://github.com/openwrt/openwrt/commit/c8c2f522625c8d9eb5acad317e3de7a6b156decf), please rebase this PR on top of master again.

Done rebasing. Thank you :)

darknblack commented:
Hi @dangowrt, what's the next step?

darknblack commented:
Hi @dangowrt @csharper2005 just want to follow up on this. What should be the next step? Thank you

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,272 @@
+			linux,input-type = <EV_SW>;
+			gpios = <&pio 0 GPIO_ACTIVE_HIGH>;
+		};
+		
```
Here are some strange blank characters. Please remove them.

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,272 @@
+		};
+
+		led-status-blue {
+			label = "blue:status";
```
All led labels can be removed now after rebasing to master branch.

darknblack replied:
Removed. Thanks 🤞

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,272 @@
+			function = LED_FUNCTION_STATUS;
+			gpios = <&pio 23 GPIO_ACTIVE_HIGH>;
+		};
+
```
Here are some strange blank characters. Please remove them.

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,272 @@
+		wlan2g-green {
+			label = "green:wlan2g";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WLAN;
```
Request extra `function-enumerator = <0>;`.

raenye replied:
Why `<0>` and `<1>`? @chunkeey once suggested `<2>` and `<5>`, and @csharper2005 (IIRC) used `<24>` and `<50>` somewhere.

DragonBluep replied:
It's just my preference. I prefer using {0,1,2} to match the radio{0,1,2} and phy{0,1,2} displayed in Luci and iwinfo. Some devices have multiple 5GHz phys. And some phys can even support both 2G and 5G. So just choose what you like.

darknblack replied:
- Added function-enumerator = <0>
- Removed label
- Renamed the led id from `wlan2g-green` to `led-wlan2g` and `wlan5g-green` to `led-wlan5g`

Thanks @raenye and @DragonBluep 🤞

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,272 @@
+		wlan5g-green {
+			label = "green:wlan5g";
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WLAN;
```
Request extra `function-enumerator = <1>;`.

darknblack replied:
- Added function-enumerator = <1>
Thanks @DragonBluep 🤞

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,272 @@
+			pause;
+		};
+	};
+	
```
Here are some strange blank characters. Please remove them.

DragonBluep reviewed (changes requested):
Please use Linux new line style `LF` instead of the Windows style `CRLF` in the new dts file.

csharper2005 commented:
> Hi @dangowrt @csharper2005 just want to follow up on this. What should be the next step? Thank you

@darknblack Hi! I suggest comment and resolve all active conversations and then wait until one of the commiters (with write access) merge the PR. It can take a while. It would also be nice if other device owners make tests and leaves their `Tested-By: xxx xxxx <xxxx@xxx.xx>` here.

darknblack committed:
mediatek: add support for Zbtlink ZBT-Z8103AX

Specifications:
SoC: MediaTek MT7981B
RAM: 256MiB
Flash: SPI-NAND 128 MiB
Switch: 1 WAN, 3 LAN (Gigabit)
Buttons: Reset, Mesh
Power: DC 12V 1A
WiFi: MT7976CN
UART: 115200n8
UART Layout:
VCC-RX-TX-GND

No. of Antennas: 6
Note: Upon opening the router, only 5 antennas were connected
to the mainboard.

Led Layout:
Power-Mesh-5gwifi-WAN-LAN3-LAN2-LAN1-2gWiFi

Buttons:
Reset-Mesh

Installation:
A. Through OpenWrt Dashboard:
If your router comes with OpenWrt preinstalled (modified by the seller),
you can easily upgrade by going to the dashboard (192.168.1.1) and then
navigate to System -> Backup/Flash firmware, then flash the firmware

B. Through TFTP
Standard installation via UART:

1. Connect USB Serial Adapter to the UART, (NOTE: Don't connect the VCC pin).
2. Power on the router. Make sure that you can access your router via UART.
3. Restart the router then repeatedly press ctrl + c to skip default boot.
4. Type > bootmenu
5. Press '2' to select upgrade firmware
6. Press 'Y' on 'Run image after upgrading?'
7. Press '0' and hit 'enter' to select TFTP client (default)
8. Fill the U-Boot's IP address and TFTP server's IP address.
9. Finally, enter the 'firmware' filename.

Signed-off-by: Ian Oderon <ianoderon@gmail.com>

darknblack commented:
Hi, I resolved all active conversations. 

Changes:
- Removed weird characters (CRLF to LF).
- Removed the led labels
- Renamed the led id of wlan2g and wlan5g.
- Added `function-enumator` for `led-wlan2g` and `led-wlan5g`.

I installed the firmware on my router and it worked perfectly: 
- LAN ports
- LED lights
- WLAN (2g & 5g)
- MAC address

Hopefully we can merge this to main branch soon. Thank you.

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,269 @@
+		conf-pu {
+			pins = "SPI0_CS", "SPI0_HOLD", "SPI0_WP";
+			drive-strength = <8>;
+			bias-pull-up = <103>;
```
Please replace `8` with `MTK_DRIVE_8mA` and replace `103` with `MTK_PUPD_SET_R1R0_11`.

DragonBluep commented on target/linux/mediatek/dts/mt7981b-zbtlink-zbt-z8103ax.dts:
```diff
@@ -0,0 +1,269 @@
+		conf-pd {
+			pins = "SPI0_CLK", "SPI0_MOSI", "SPI0_MISO";
+			drive-strength = <8>;
+			bias-pull-down = <103>;
```
Same change here.

darknblack replied:
Replaced. Thank you 🙆‍♂️ 🤞

DragonBluep reviewed (commented):
Hi! I found some small improvements.

darknblack commented:
Hi any updates? Any reviewers/contributors?
